packages/web-api/src/file-upload.ts

```diff
@@ -0,0 +1,315 @@
+import { Logger } from '@slack/logger';
+import { readFileSync, ReadStream } from 'fs';
+import { errorWithCode, ErrorCode } from './errors';
+import { FilesCompleteUploadExternalArguments, FilesUploadV2Arguments, FileUploadV2, FileUploadV2Entry } from './methods';
+
+/**
+ * Returns a fileUploadEntry used internally to store state of the file upload
+ * @param options Options provided by user
+ * @param channelId optional channel id to share file with, omitted, channel is private
+ * @returns
+*/
+export async function getFileUpload(
+  options: FilesUploadV2Arguments | FileUploadV2,
+  logger: Logger, channelId?: string,
+): Promise<FileUploadV2Entry> {
+  warnIfLegacyFileType(options, logger);
+  const fileName = warnIfMissingOrInvalidFileNameAndDefault(options, logger);
+  const fileData = await getFileData(options);
+  const fileDataBytesLength = getFileDataLength(fileData);
+
+  const fileUploadEntry = {
+    // supplied by user
+    alt_text: options.alt_text,
+    channel_id: channelId,
+    content: options.content,
+    file: options.file,
+    filename: options.filename ?? fileName,
+    initial_comment: options.initial_comment,
+    snippet_type: options.snippet_type,
+    thread_ts: options.thread_ts,
+    title: options.title ?? (options.filename ?? fileName),
+    // calculated
+    data: fileData,
+    length: fileDataBytesLength,
+  };
+  return fileUploadEntry;
+}
+
+/**
+ * Returns an array of files upload entries when `file_uploads` is supplied.
+ * **Note**
+ * file_uploads can be set as options that are intended to be attached to a
+ * single message. To support this, we will handle options supplied either
+ * in both following patterns:
+ * ### Pattern 1
+ * Top level `initial_comment`, `thread_ts`, `channel_id`
+ * ```javascript
+ * const res = await client.files.uploadV2({
+ *   initial_comment: 'Here are the files!',
+ *   thread_ts: '1223313423434.131321',
+ *   channel_id: 'C123',
+ *   file_uploads: [
+ *     {
+ *       file: './test/fixtures/test-txt.txt',
+ *       filename: 'test-txt.txt',
+ *     },
+ *     {
+ *       file: './test/fixtures/test-png.png',
+ *       filename: 'test-png.png',
+ *     },
+ *   ],
+ * });
+ * ```
+ * ### Pattern 2
```

Comment:
I don't recommend supporting this pattern because of the above reason. Also, we don't support this way in Python and Java SDKs now.